app/internal/import_file.py

```diff
@@ -0,0 +1,199 @@
+    """
+    same_date_counter = 1
+    date_n_count = defaultdict(int)
+    for event in events:
+        date_n_count[event["S_Date"]] += 1
+        if date_n_count[event["S_Date"]] > same_date_counter:
+            same_date_counter = date_n_count[event["S_Date"]]
+    return same_date_counter <= max_event_start_date
+
+
+def open_txt_file(txt_file: str) -> List[str]:
+    with open(txt_file, "r") as text:
+        return text.readlines()
+
+
+def save_calendar_content_txt(event: str, calendar_content: List) -> bool:
+    """populate calendar with event content"""
+    head, content, start_date, end_date = event.split(", ")
+    if (not is_date_in_range(start_date) or
+       not is_date_in_range(end_date.replace("\n", ""))):
+        return False
+    start_date = datetime.datetime.strptime(start_date, DATE_FORMAT)
+    end_date = datetime.datetime.strptime(end_date.replace("\n", ""),
+                                          DATE_FORMAT)
+    calendar_content.append({"Head": head,
+                             "Content": content,
+                             "S_Date": start_date,
+                             "E_Date": end_date})
+    return True
+
+
+def import_txt_file(txt_file: str) -> List[Dict[str, Union[str, Any]]]:
+    calendar_content = []
+    events = open_txt_file(txt_file)
+    for event in events:
+        if (not is_event_text_valid(event) or
+           not save_calendar_content_txt(event, calendar_content)):
+            return []
+    return calendar_content
+
+
+def open_ics(ics_file: str) -> Union[List, Calendar]:
+    with open(ics_file, "r") as ics:
+        try:
+            calendar_read = Calendar.from_ical(ics.read())
+        except (IndexError, ValueError):
+            return []
+    return calendar_read
+
+
+def is_valid_data_event_ics(component) -> bool:
+    """check if ics event data content is valid"""
+    return not (str(component.get('summary')) is None or
+                component.get('dtstart') is None or
+                component.get('dtend') is None or
+                not is_date_in_range(component.get('dtstart').dt) or
+                not is_date_in_range(component.get('dtend').dt))
+
+
+def save_calendar_content_ics(component, calendar_content) -> None:
+    calendar_content.append({
+                        "Head": str(component.get('summary')),
+                        "Content": str(component.get('description')),
+                        "S_Date": component.get('dtstart').dt
+                        .replace(tzinfo=None),
+                        "E_Date": component.get('dtend').dt
+                        .replace(tzinfo=None)
+                    })
+
+
+def import_ics_file(ics_file: str) -> List[Dict[str, Union[str, Any]]]:
+    calendar_content = []
+    calendar_read = open_ics(ics_file)
+    if not calendar_read:
+        return []
+    for component in calendar_read.walk():
+        if component.name == DESC_EVENT:
+            if is_valid_data_event_ics(component):
```

Comment:
Flip the condition to return early

Reply:
the first `if `checks if the event is "VEVENT", other events like "VTODO" and "VJOURNAL" not interest us and we skip them.
the nested `if `checks if the "VEVENT" fields are valid. so I'm thinking to keep it in this order if it is ok.

Reply:
My intention was to rewrite the code like that:

```python
            if not is_valid_data_event_ics(component):
                return []
            save_calendar_content_ics(component, calendar_content)
```

Is there any logical difference that I miss here?

Reply:
I thought you want to change the if's order my bad :)
I changed it.
In addition, after you accept this ticket could I open a new one after I characterize all the new changes required and tests?
(I want to add location support and more start and end date tests for our code)

Reply:
Yeah sure why not :)